src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs

```diff
@@ -144,29 +207,44 @@ public void ProcessRequests()
 
                         try
                         {
-                            this.testCaseEventMonitorTask.Wait();
+                            this.testCaseEventMonitorTask.Wait(this.cancellationTokenSource.Token);
+                            this.dataCollectionTestCaseEventHandler.Close();
                         }
                         catch (Exception ex)
                         {
-                            EqtTrace.Error("DataCollectionRequestHandler.ProcessRequests : {0}", ex.Message);
+                            if (EqtTrace.IsErrorEnabled)
+                            {
+                                EqtTrace.Error("DataCollectionRequestHandler.ProcessRequests : {0}", ex.Message);
+                            }
                         }
 
                         var attachments = this.dataCollectionManager.SessionEnded(isCancelled);
```

Comment:
Suppose this is a heavy operation, the message loop will not process any other messages (including cancel) until this is complete.